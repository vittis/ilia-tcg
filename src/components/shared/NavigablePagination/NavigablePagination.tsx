import {
	Pagination,
	PaginationContent,
	PaginationEllipsis,
	PaginationItem,
	PaginationButton,
	PaginationNext,
	PaginationPrevious,
} from "@/components/ui/pagination";
import { useEffect, useMemo, useRef } from "react";
import { useSearchParams } from "react-router-dom";

interface NavigablePaginationProps {
	currentPage: number;
	isLoading: boolean;
	totalCount?: number;
	pageSize?: number;
}

const NavigablePagination = ({
	currentPage,
	totalCount = 0,
	pageSize = 0,
	isLoading,
}: NavigablePaginationProps) => {
	let [_, setSearchParams] = useSearchParams();

	/* Needed to maintain last total pages when search query changes */
	const lastTotalPagesRef = useRef(0);

	const totalPages = totalCount ? Math.ceil(totalCount / pageSize) : lastTotalPagesRef.current;

	useEffect(() => {
		lastTotalPagesRef.current = totalPages;
	}, [totalPages]);

	const paginationItems = useMemo(() => {
		if (totalPages <= 5) {
			return Array.from({ length: totalPages }, (_, i) => (
				<PaginationItem key={i + 1}>
					<PaginationButton
						disabled={isLoading}
						isActive={currentPage === i + 1}
						onClick={() => setSearchParams({ page: (i + 1).toString() })}
					>
						{i + 1}
					</PaginationButton>
				</PaginationItem>
			));
		} else {
			const middlePages = Array.from(
				{ length: Math.min(totalPages - 2, 3) },
				(_, i) => currentPage - 1 + i,
			).filter(page => page > 1 && page < totalPages);

			return [
				<PaginationItem key={1}>
					<PaginationButton
						disabled={isLoading}
						isActive={currentPage === 1}
						onClick={() => setSearchParams({ page: "1" })}
					>
						1
					</PaginationButton>
				</PaginationItem>,
				currentPage > 3 && (
					<PaginationItem key="ellipsis1">
						<PaginationEllipsis />
					</PaginationItem>
				),
				...middlePages.map(page => (
					<PaginationItem key={page}>
						<PaginationButton
							disabled={isLoading}
							isActive={currentPage === page}
							onClick={() => setSearchParams({ page: page.toString() })}
						>
							{page}
						</PaginationButton>
					</PaginationItem>
				)),
				currentPage < totalPages - 2 && (
					<PaginationItem key="ellipsis2">
						<PaginationEllipsis />
					</PaginationItem>
				),
				<PaginationItem key={totalPages}>
					<PaginationButton
						disabled={isLoading}
						isActive={currentPage === totalPages}
						onClick={() => setSearchParams({ page: totalPages.toString() })}
					>
						{totalPages}
					</PaginationButton>
				</PaginationItem>,
			].filter(Boolean);
		}
	}, [currentPage, isLoading, totalPages, setSearchParams]);

	return (
		<Pagination className="mt-10">
			<PaginationContent>
				<PaginationItem>
					<PaginationPrevious
						disabled={currentPage === 1 || isLoading}
						onClick={() => setSearchParams({ page: `${currentPage - 1}` })}
					/>
				</PaginationItem>

				{paginationItems}

				<PaginationItem>
					<PaginationNext
						disabled={currentPage === totalPages || isLoading}
						onClick={() => setSearchParams({ page: `${currentPage + 1}` })}
					/>
				</PaginationItem>
			</PaginationContent>
		</Pagination>
	);
};

export { NavigablePagination };
