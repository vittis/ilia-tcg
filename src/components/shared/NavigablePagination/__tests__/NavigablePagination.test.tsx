import { render, screen } from "@testing-library/react";
import { NavigablePagination } from "../NavigablePagination";
import { BrowserRouter } from "react-router-dom";

describe("NavigablePaginationigablePagination", () => {
	it("matches snapshot", () => {
		const { container } = render(
			<BrowserRouter>
				<NavigablePagination currentPage={1} isLoading={false} pageSize={250} totalCount={2500} />,
			</BrowserRouter>,
		);

		expect(container).toMatchSnapshot();
	});

	it("should render correct pages number of when currentPage=1", () => {
		render(
			<BrowserRouter>
				<NavigablePagination currentPage={1} isLoading={false} pageSize={250} totalCount={2500} />
			</BrowserRouter>,
		);

		expect(screen.getAllByRole("button")).toHaveLength(5);
	});

	it("should render correct number of pages when currentPage is in middle", () => {
		render(
			<BrowserRouter>
				<NavigablePagination currentPage={8} isLoading={false} pageSize={250} totalCount={2500} />
			</BrowserRouter>,
		);

		expect(screen.getAllByRole("button")).toHaveLength(7);
	});

	it("should render correct number of pages when currentPage is last", () => {
		render(
			<BrowserRouter>
				<NavigablePagination currentPage={10} isLoading={false} pageSize={250} totalCount={2500} />
			</BrowserRouter>,
		);

		expect(screen.getAllByRole("button")).toHaveLength(5);
	});
});
