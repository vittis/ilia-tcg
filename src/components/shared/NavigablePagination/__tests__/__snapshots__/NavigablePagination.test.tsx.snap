// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigablePaginationigablePagination > matches snapshot 1`] = `
<div>
  <nav
    aria-label="pagination"
    class="mx-auto flex w-full justify-center mt-2"
    role="navigation"
  >
    <ul
      class="flex flex-row items-center gap-1"
    >
      <li
        class=""
      >
        <button
          aria-label="Go to previous page"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 gap-1 pl-2.5"
          disabled=""
        >
          <svg
            class="lucide lucide-chevron-left h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m15 18-6-6 6-6"
            />
          </svg>
          <span>
            Previous
          </span>
        </button>
      </li>
      <li
        class=""
      >
        <button
          aria-current="page"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10"
        >
          1
        </button>
      </li>
      <li
        class=""
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10"
        >
          2
        </button>
      </li>
      <li
        class=""
      >
        <span
          aria-hidden="true"
          class="flex h-9 w-9 items-center justify-center"
        >
          <svg
            class="lucide lucide-ellipsis h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="1"
            />
            <circle
              cx="19"
              cy="12"
              r="1"
            />
            <circle
              cx="5"
              cy="12"
              r="1"
            />
          </svg>
          <span
            class="sr-only"
          >
            More pages
          </span>
        </span>
      </li>
      <li
        class=""
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 w-10"
        >
          10
        </button>
      </li>
      <li
        class=""
      >
        <button
          aria-label="Go to next page"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 gap-1 pr-2.5"
        >
          <span>
            Next
          </span>
          <svg
            class="lucide lucide-chevron-right h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
        </button>
      </li>
    </ul>
  </nav>
  ,
</div>
`;
